core/transformer/src/IModelImporter.ts

```diff
@@ -130,14 +135,25 @@ export class IModelImporter {
 
   /** Import the specified ElementProps (either as an insert or an update) into the target iModel. */
   public importElement(elementProps: ElementProps): Id64String {
-    if (undefined !== elementProps.id) {
-      if (this.doNotUpdateElementIds.has(elementProps.id)) {
-        Logger.logInfo(loggerCategory, `Do not update target element ${elementProps.id}`);
-        return elementProps.id;
+    if (undefined !== elementProps.id && this.doNotUpdateElementIds.has(elementProps.id)) {
+      Logger.logInfo(loggerCategory, `Do not update target element ${elementProps.id}`);
+      return elementProps.id;
+    }
+    if (this.preserveElementIdsForFiltering) {
+      // categories are the only element that onInserted will immediately insert a new element (their default subcategory)
+      // since default subcategories always exist and always will be inserted after their categories, we treat them as an update
+      // to prevent duplicate inserts
+      if (isSubCategory(elementProps) && isDefaultSubCategory(elementProps)) {
+        this.onUpdateElement(elementProps);
```

Comment:
It seems unlikely given the nature of subcategories but the way you refactored handling undefined elementProps.id it could technically be null when you make this call

Reply:
Looking at the implementation of `IModelDb.getDefaultSubCategoryId`, I think `isDefaultSubCategory` may indeed allow this through when elementProps.id is the invalid id, I'll add an explicit check to both functions used in this branch

Reply:
updated, please take a look